src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs

```diff
@@ -1306,6 +1310,32 @@ internal class PipelineStopper
         /// Object used for synchronization.
         /// </summary>
         private object _syncRoot = new object();
+
+        internal void AwaitPipelineCriticalSection()
+        {
+            _criticalSectionSemaphore.Wait();
+            _criticalSectionSemaphore.Release();
+        }
+
+        internal void EnterCriticalSection()
+        {
+            Interlocked.Increment(ref _criticalSectionDepth);
+            _criticalSectionSemaphore.Wait(millisecondsTimeout: 0);
+        }
+
+        internal void ExitCriticalSection()
+        {
+            if (Interlocked.Decrement(ref _criticalSectionDepth) == 0)
+            {
+                _criticalSectionSemaphore.Release();
+            }
+        }
```

Comment:
No, the depth should be able to be incremented even if it's not-zero. This could be called successively (a function calls another function from its cleanup block, and that called function also has a cleanup block) without exiting the entirety of the critical region.

Reply:
> No, the depth should be able to be incremented even if it's not-zero

Yeah, my example does that. `Interlocked.Increment(ref _criticalSectionDepth)` will always increment the field passed in.

In

```csharp
Interlocked.Increment(ref _criticalSectionDepth);
_criticalSectionSemaphore.Wait(0);
```

Every thread increments the counter and *every* thread tries to take out the semaphore.

In

```csharp
if (Interlocked.Increment(ref _criticalSectionDepth) == 0)
{
    _criticalSectionSemaphore.Wait();
}
```

Every thread increments the counter, but *only the first thread* takes out the semaphore, because only its return result is 0.

In the case of using semaphores like this, there's no actual difference, because `Wait(0)` continues on whether or not it succeeded. But we already have the information we need to determine whether or not to actually perform that call.

Reply:
Ah okay, I see what you're getting at there. I think I'm gonna have to add a comment explaining what that's going to do as well, if only for my own sanity 😂

Reply:
[Docs](https://docs.microsoft.com/en-us/dotnet/api/system.threading.interlocked.increment?view=netcore-3.1) suggest that the returned value is the value after the increment is applied, so I guess we should check it comes out to 1 instead.